File: Problem2.cpp
// Min Heap

// Time Complexity: Insert, heapify and remove - O(logN) 
// Space Complexity: O(1)
// Any problem you faced while coding this : no.


// Your code here along with comments explaining your approach
// Using array to store the heap elements. We check everytime we push an element if its the minimum of all the nodes 
// present in the tree. When deleting an element, get the last element to the first location in the vector
// (and remove that last element once copied) and we do heapify operation to arrange elements in order.

#include<vector>
#include<iostream>

class MinHeap {
    private:
    int capacity;
    int size;
    std::vector<int> heap;
    
    // Helper to swap two nodes of the heap.
    void swap(int fpos, int spos)
    {
        int temp;
        temp = heap[fpos];
        heap[fpos] = heap[spos];
        heap[spos] = temp;
        return;
    }

    public: 
    // constructor
    MinHeap(int cap)
        : capacity(cap) 
    {
        heap.resize(capacity+1);
        heap[0] = INT32_MIN;
    } 

    // Destructor    
    ~MinHeap() {}

    //  Get size
    int getSize(void)
    {
        return size;
    }

    // Get Parent node of the given node
    int getParent(int pos)
    {
        return pos/2;
    }

    // Get LeftChild of the given node
    int getLeftChild(int pos)
    {
        return (2*pos);
    }

    // Get RightChild of the given node
    int getRightChild(int pos)
    {
        return ((2*pos) + 1);
    }

    // Check if the given node is a leaf node
    bool isLeafNode(int pos)
    {
        if((pos > size/2))
        {
            return true;
        } 
        return false;
    }

    // Insert the element into appropriate place.
    void insert(int element)
    {
        if(size >= capacity)
        {
            return;
        }
        heap[size] = element;
        int curr = size++;
        int parentIdx = getParent(curr);
        while(curr > 0 && (heap[curr] < heap[parentIdx]))
        {
            swap(curr, getParent(curr));
            curr = parentIdx;
            parentIdx = getParent(curr);
        }
        return;
    }

    // Helper to place elements according to the sorting order.
    void heapify(int index)
    {
        if(!isLeafNode(index))
        {
            int leftChildIdx = getLeftChild(index);
            int rightChildIdx = getRightChild(index);
            int smallestChildIndex = leftChildIdx;
            if((rightChildIdx < size) && (heap[rightChildIdx] < heap[leftChildIdx]))
            {
                smallestChildIndex = rightChildIdx;
            }
            
            if((heap[index] > heap[leftChildIdx]) || (heap[index] > heap[rightChildIdx]))
            {
                swap(index, smallestChildIndex);
                heapify(smallestChildIndex);
            }
        }
        return;
    }

    // Remove the first element
    int remove(void)
    {
        int removedElem = heap[1];
        heap[1] = heap[size--];
        heapify(1);
        return removedElem;
    }

    void printElements(void)
    {
        std::cout << "elements are: " << std::endl;
        for(int i = 1; i < size/2; i++)
        {
            std::cout << "Parent: " << heap[i]  << "\t" << "LeftChild" << heap[2*i] << "\t" << "RightChild" << heap[2*i+1] << std::endl;
        }
        return;
    }  
};

int main()
{
    MinHeap minHeap(15);

    // Custom input entries
    minHeap.insert(5);
    minHeap.insert(3);
    minHeap.insert(17);
    minHeap.insert(10);
    minHeap.insert(84);
    minHeap.insert(19);
    minHeap.insert(6);
    minHeap.insert(22);
    minHeap.insert(9);

    // Print all elements of the heap
    minHeap.printElements();

    // Delete first element
    std::cout << "min val - " << minHeap.remove() << std::endl;

    return 0;
}

File: Problem2.py
# Min Heap
# Time Complexity
# getMin - O(1), Insert, Delete, Heapify - O(log(n))

class MinHeap:

    def __init__(self):
        self.heap = []

    def getMin(self):
        if len(self.heap) == 0:
            return float('-inf')
        return self.heap[0]

    def insert(self, val):
        self.heap.append(val)
        curIdx = len(self.heap) - 1 #length will be at least 1

        while curIdx > 0 and self.getParent(curIdx) > self.heap[curIdx]:
            self.swap(curIdx, self.getParentIdx(curIdx))
            curIdx = self.getParentIdx(curIdx)
        
        return curIdx

    def deleteMin(self):
        if len(self.heap) == 0:
            return float('-inf')
        else:
            self.swap(0, len(self.heap) - 1)
            self.heap.pop()
            self.heapify(0)

    def deleteIndex(self, index):
        if index < len(self.heap):
            self.heap[index] = self.getMin() - 1

        curIdx = index
        while curIdx > 0 and self.getParent(curIdx) > self.heap[curIdx]:
            self.swap(curIdx, self.getParentIdx(curIdx))
            curIdx = self.getParentIdx(curIdx)
        
        self.deleteMin()
        

    #considering valid calls only
    def heapify(self, index):
        smallestIdx = index

        if self.getLeftChildIdx(index) < len(self.heap) and self.getLeftChild(index) < self.heap[smallestIdx]:
            smallestIdx = self.getLeftChildIdx(index)

        if self.getRightChildIdx(index) < len(self.heap) and self.getRightChild(index) < self.heap[smallestIdx]:
            smallestIdx = self.getRightChildIdx(index)
        
        if smallestIdx != index:
            self.swap(index, smallestIdx)
            self.heapify(smallestIdx)
        



# TC
# 0, heap = [0], curIdx = 0
# -1, heap = [0, -1], curIdx = 1, parentIdx = (1-1)//2 = 0, swap!, heap = [-1 , 0]
# 2, heap = [-1, 0, 2]

    




    #helper methods

    #considering only valid calls
    def swap(self, i1, i2):
        temp = self.heap[i1]
        self.heap[i1] = self.heap[i2]
        self.heap[i2] = temp
    
    def getLeftChildIdx(self, index):
        return 2*index + 1

    def getRightChildIdx(self, index):
        return 2*index + 2

    def getParentIdx(self, index):
        return (index - 1)//2

    #considering only valid calls
    def getLeftChild(self, index):
        return self.heap[self.getLeftChildIdx(index)]

    #considering only valid calls
    def getRightChild(self, index):
        return self.heap[self.getRightChildIdx(index)]

    #considering only valid calls
    def getParent(self, index):
        return self.heap[self.getParentIdx(index)]



def main():
    minHeap = MinHeap()
    minHeap.insert(9)
    minHeap.insert(8)
    minHeap.insert(7)
    minHeap.insert(6)
    minHeap.insert(5)
    minHeap.insert(4)
    minHeap.insert(3)
    print(minHeap.getMin())
    print(minHeap.heap)
    minHeap.deleteMin()
    print(minHeap.heap)

main()
    